modules/auxiliary/scanner/ssl/ssl_version.rb

```diff
@@ -0,0 +1,519 @@
+    # such as :TLSv1, :TLSv1_client, :TLSv1_server. In this case, we only need :TLSv1, so we need to remove ~2/3 of the list.
+    # Next, we'll find that many ciphers in OpenSSL::Cipher.ciphers are not applicable for various SSL versions.
+    # The loop we previously looked at has (at the time of writing on Kali Rollin, msf 6.2.23) 3060 rounds.
+    # This is a lot of iterations when we already know there are many combinations that will not be applicable for our
+    # use.
+    # Luckily there is a 2nd way which is much more efficent.
+    # The OpenSSL library includes https://docs.ruby-lang.org/en/2.4.0/OpenSSL/SSL/SSLContext.html#method-i-ciphers
+    # which we can use to generate a list of all ciphers, and SSL versions they work with. The structure is:
+    # [[name, version, bits, alg_bits], ...]
+    # which makes it very easy to just pull the 2nd element (version, or SSL version) from each list item, and unique it.
+    # This gives us the list of all SSL versions which also have at least one working cipher on our system.
+    # Using this method we produce no unusable SSL versions or matching cipher suites and the list is 60 items long, so 1/51 the size.
+    # Later in get_metasploit_ssl_cipher_suites, we can grab all cipher suites to a SSL version easily by simply filtering
+    # the 2nd element (version, or SSL version) from each list item.
+
+    if datastore['SSLVersion'] == 'All'
+      return Array.new(OpenSSL::SSL::SSLContext.new.ciphers.length) { |i| (OpenSSL::SSL::SSLContext.new.ciphers[i][1]).to_s }.uniq.reverse
+    end
+
+    datastore['SSLVersion']
+  end
+
+  def get_metasploit_ssl_cipher_suites(ssl_version)
+    # See comments in get_metasploit_ssl_versions for details on the use of
+    # OpenSSL::SSL::SSLContext.new.ciphers verse other methods to generate
+    # valid ciphers for a given SSL version
+
+    # First find all valid ciphers that the Metasploit host supports.
+    # Also transform the SSL version to a standard format..
+    ssl_version = ssl_version.to_s.gsub('_', '.')
+    all_ciphers = OpenSSL::SSL::SSLContext.new.ciphers
+    valid_ciphers = []
+
+    # For each cipher that the Metasploit host supports, determine if that cipher
+    # is supported for use with the SSL version passed into this function.
+    all_ciphers.each do |cipher|
+      # cipher list has struct of [cipher, ssl_version, <int>, <int>]
+      if cipher[1] == ssl_version
+        valid_ciphers << cipher[0]
+      end
+    end
+
+    # If the user wants to use all ciphers then return all valid ciphers.
+    # Otherwise return only those that match the ones the user specified
+    # in the SSLCipher datastore option.
+    #
+    # If no match is found for some reason then we will return an empty array.
+    if datastore['SSLCipher'] == 'All'
+      return valid_ciphers
+    elsif valid_ciphers.contains? datastore['SSLCipher']
+      return [datastore['SSLCipher']]
+    end
```

Comment:
Question but this is assuming the user specifies a single value yet your description says "match the ones the user specified" which suggests plural.

Therefore not sure if this part of the code is meant to be returning one value or multiple? Given your description says that this contains the ciphers to test I would believe that you could specify multiple and not just alternate between ALL and a specific cipher.

Reply:
you are right, its a verbiage problem, it should be singular